### PR TITLE
Add Black Friday page

### LINKED ISSUE #155: Add Black Friday page

### PR DESCRIPTION
**Pull request by wkenned1**
Fixes #155

- [x] black background
- [x] white background and black text for cards
- [x] orange new price to make it distinctive
- [x] new price setting (in the code) (@daniel-ciocirlan)
- [x] course coupon links on the course platform (@daniel-ciocirlan)
- [x] black/white logo and Black Friday bundle card (@daniel-ciocirlan)
- [x] new Black Friday product page (@daniel-ciocirlan)
- [x] remove the free courses
- [x] reorder courses: rites of passage first, bundles next, regular courses after
- [x] collapsible Black Friday banner on blog posts
- [x] replace bundle image
- [x] have bundle go directly to purchase page
- [x] add a learn more button to go to the black friday course page
- [x] make course link go to purchase page with automatic discount applied
- [x] Add discount code to banner (@daniel-ciocirlan)
- [x] Add a learn more and checkout option for courses
- [x] Add css to price and make it stand out more
- [x] Move old course back into regular groupings
- [x] Still free category
- [x] Add newsletter component at the bottom of the list
- [x] Rename RoP to project based courses
- [x] Change CSS for hero component
- [x] Change hero section messages
- [x] Add description (@daniel-ciocirlan)
- [x] Add subtitle1 (@daniel-ciocirlan)
- [x] Fix course card edges (@wkenned1)
- [x] Remove banner JS and X
- [x] Center text on banner (@wkenned1)
- [x] Add classic akka discount prices (@daniel-ciocirlan)
- [x] Make sure newsletter form upload works

### CODE REVIEW

**Comment by daniel-ciocirlan** [comment]
TODO after our conversation

- [ ] black background
- [ ] white background and black text for cards
- [ ] orange new price to make it distinctive
- [ ] new price setting (in the code) (@daniel-ciocirlan)
- [ ] course coupon links on the course platform (@daniel-ciocirlan)
- [ ] black/white logo and Black Friday bundle card (@daniel-ciocirlan)
- [ ] new Black Friday product page (@daniel-ciocirlan)
- [ ] remove the free courses
- [ ] reorder courses: rites of passage first, bundles next, regular courses after
- [ ] collapsible Black Friday banner on blog posts

Let me know if I've missed anything
Let me know if I've missed anything

**Comment by daniel-ciocirlan** [comment]
Hey @andreilgeorgescu @wkenned1 I've pushed the BF product page on your branch, should not conflict with other files. I've left some TODOs on the product page so that I can add minimal descriptions later (my 2 functioning neurons at the moment are not very creative), and will push the last changes (and any minimal nits I find) over the weekend.

I think I/we can merge this PR on Sunday before the sale starts.

**Comment by andreilgeorgescu** [comment]
When the banner is closed, it currently does not stay closed on other pages.

Personally, I think that there doesn't need to be a way to close the banner because I don't find it intrusive, it adds a seasonal character to the site, and it is not sticky.

What do you think @daniel-ciocirlan?

**Comment by daniel-ciocirlan** [comment]
@wkenned1 fine not to close the banner. I'll continue with the rest of the tasks over the weekend

**Comment by andreilgeorgescu** [comment]
> @andreilgeorgescu fine not to close the banner. I'll continue with the rest of the tasks over the weekend

Sounds good. I added a few rough experimental ideas there for you to see revolving around CSS. And we put the newsletter form after the bundle section to see what it looks like there.

File: src/components/CourseEntry.astro [hunk 1]
```diff
@@ -11,9 +11,9 @@ interface Props {
   holiday?: boolean;
 }
 
-const { course, holiday = false, prices } = Astro.props;
+const { course, holiday = false, prices, productId, coupon } = Astro.props;
```
Comment (andreilgeorgescu) [inline src/components/CourseEntry.astro:14]: Should be in interface.

File: src/components/CourseCard.astro [hunk 2]
```diff
@@ -21,16 +21,30 @@ const { course, holiday = false, prices } = Astro.props;
...
+    <div class="relative h-72 w-full overflow-hidden md:h-full md:w-64 lg:w-96">
+      <a href="href={`/courses/${course.slug}`}">
         <Picture
           src={course.data.heroImage}
           alt={`${course.data.title}`}
           class="min-h-full min-w-full object-cover object-center"
           formats={["avif", "webp"]}
         />
+      </a>
+      <div class="absolute bottom-0 flex w-full">
```
Comment (andreilgeorgescu) [inline src/components/CourseCard.astro:33]: This course card component is used on the frontpage of the site so anything specific to the holiday card should be behind conditional checks as to not affect non-holiday content.

File: src/layouts/HolidayLayout/_components/HolidayDiscountBundle.astro [hunk 3]
```diff
@@ -9,20 +9,24 @@ interface Props {
...
-        class="mx-auto flex max-w-2xl flex-col gap-16 px-6 sm:rounded-3xl sm:p-8 lg:mx-0 lg:max-w-none lg:flex-row lg:items-center xl:gap-x-20 xl:px-20"
+        class="mx-auto flex max-w-2xl flex-col gap-16 px-6 sm:p-8 lg:mx-0 lg:max-w-none lg:flex-row lg:items-center xl:gap-x-20 xl:px-20"
       >
-        <Picture
-          src={bundleImage}
-          alt={`${title} bundle`}
-          class="w-full flex-none lg:max-w-lg"
-          formats={["avif", "webp"]}
-        />
+        <a href={checkoutURL}>
```
Comment (andreilgeorgescu) [inline src/layouts/HolidayLayout/_components/HolidayDiscountBundle.astro:22]: As a user my expectation for clicking on the image would be to take me to the "learn more" page.

**Review by andreilgeorgescu — CHANGES_REQUESTED** [review]

File: src/pages/black-friday-2024/_data/courses.ts [hunk 4]
```diff
@@ -2,21 +2,24 @@ export const _courses: {
   category: string;
   courses: {
     id: string;
+    productId: number;
```
[the remark this answers is not in the document]
Comment (daniel-ciocirlan) [reply src/pages/black-friday-2024/_data/courses.ts:5]: A good and important catch - fixed

File: src/components/CourseCard.astro [hunk 5]
```diff
@@ -12,7 +12,7 @@ interface Props {
   };
 }
 
-const { course, holiday = false, prices } = Astro.props;
+const { course, holiday = false, coupon, productId, prices } = Astro.props;
```
[the remark this answers is not in the document]
Comment (daniel-ciocirlan) [reply src/components/CourseCard.astro:15]: fixed

File: src/components/CourseEntry.astro (hunk 1 above)
Comment (daniel-ciocirlan) [reply src/components/CourseEntry.astro:14]: fixed

File: src/components/CourseCard.astro (hunk 2 above)
Comment (daniel-ciocirlan) [reply src/components/CourseCard.astro:33]: Correct. Added checks and moved the links to their own component

File: src/layouts/HolidayLayout/_components/HolidayDiscountBundle.astro (hunk 3 above)
Comment (daniel-ciocirlan) [reply src/layouts/HolidayLayout/_components/HolidayDiscountBundle.astro]: Done

**Comment by andreilgeorgescu** [comment]
@daniel-ciocirlan [I added the classic courses to the course list.](https://github.com/rockthejvm/rockthejvm-site/pull/157/commits/8ca8a57d4a290a51043e3b26bb2369ee19529df7) If you don't want them to be shown on the sale page then we can filter them out. However, I think it's worth including for the dynamic price calculation which I added [here](https://github.com/rockthejvm/rockthejvm-site/pull/157/commits/73fd15480f930dd5fc1aeb3ee10a858d7a5d9801). I added a task for asking for the reduced prices for them.

I tried to do dynamic price calculation using what is basically a list comprehension/do notation. Can this be done in a more elegant manner?

**Comment by daniel-ciocirlan** [comment]
@andreilgeorgescu I think the dynamic price calculation is nice but complicates things for a couple of reasons:
- I would not like to show the Akka classic courses anymore
- the calculation counts bundles as well, and filtering them out would complicate things further

Also I see that the "1-click buy"/"learn more" links broke now

**Comment by andreilgeorgescu** [comment]
> @andreilgeorgescu I think the dynamic price calculation is nice but complicates things for a couple of reasons:
> 
> * I would not like to show the Akka classic courses anymore
> * the calculation counts bundles as well, and filtering them out would complicate things further
> 
> Also I see that the "1-click buy"/"learn more" links broke now

That's a good point about bundles that I forgot to consider. We could filter them out but you're right it complicates things too much for a temporary thing like this. I'll revert the change and remove the task.

I'll look into the broken link.

By the way that link looks odd on my monitor (looks fine on small width).
image

**Comment by daniel-ciocirlan** [comment]
I've made the fixes, pushing shortly @andreilgeorgescu

File: src/pages/courses/_layouts/CourseLayout.astro [hunk 6]
```diff
@@ -73,10 +73,6 @@ const response = await fetch(
   `https://rockthejvm.com/api/purchase/${pricingPlanId}`,
 );
 
-if (!response.ok) {
```
Comment (andreilgeorgescu) [inline src/pages/courses/_layouts/CourseLayout.astro:76]: What was wrong with this?

Comment (daniel-ciocirlan) [reply src/pages/courses/_layouts/CourseLayout.astro:76]: This triggers a failure on the entire page whenever the Teachable response doesn't arrive. Too aggressive, and fails in dev mode for all course pages. Kept the error just to the pricing card.

**Review by andreilgeorgescu — APPROVED** [review]

**Comment by andreilgeorgescu** [comment]
Everything seems good on our end.

One suggestion we have that you might want to consider is merging the hero and bundle discount into one CTA due to them having a level of redundancy.

image

**Pull request merged**